src/components/Bounty/bounty.utils.ts

```diff
@@ -81,3 +83,43 @@ export const getMainBountyAmount = (bounty?: {
 
   return MIN_CREATE_BOUNTY_AMOUNT;
 };
+
+export const useBountyEngagement = ({ bountyId }: { bountyId: number }) => {
+  const currentUser = useCurrentUser();
+  const queryUtils = trpc.useContext();
+
+  const { data: engagements } = trpc.user.getBountyEngagement.useQuery(undefined, {
+    enabled: !!currentUser,
+  });
+
+  const toggleEngagementMutation = trpc.user.toggleBountyEngagement.useMutation({
+    async onMutate({ type, bountyId }) {
+      await queryUtils.user.getBountyEngagement.cancel();
+      await queryUtils.bounty.getById.cancel();
+
+      const previousEngagements = queryUtils.user.getBountyEngagement.getData() ?? {};
+      const previousBounty = queryUtils.bounty.getById.getData({ id: bountyId });
+      const ids = previousEngagements[type] ?? [];
+      const isToggled = !!ids.find((id) => id === bountyId);
+
+      // TODO.bounty: optimistic update current bounty stats
+
+      queryUtils.user.getBountyEngagement.setData(undefined, (old = {}) => ({
+        ...old,
+        [type]: isToggled ? ids.filter((id) => id !== bountyId) : [...ids, bountyId],
+      }));
+
+      return { previousEngagements, previousBounty };
+    },
+    onError: (_error, _variables, context) => {
```

Comment:
Would be good to show error notification here since it'd be weird to just have it turn back